From a20ef55b0e188b82a3b65ee4cd34738c1189b31b Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Mon, 27 Oct 2025 16:41:04 +0100
Subject: [PATCH 1008/1013] evl/net: dev: ensure proper synchronization on
 device down

When a network interface goes down, we need to guarantee the following:

- the {enable, disable}_oob handlers in oob-capable NIC drivers may
  assume that no RX/TX oob traffic may take place anymore once called
  by us.

- any unconsumed/unsent skb must be released by the RX/TX threads on
  exit, making sure that no more reference on the device can be taken
  once we passed the crossing.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/dev.c    | 18 ++++++++++++++----
 kernel/evl/net/input.c  | 17 ++++++++++++++++-
 kernel/evl/net/output.c |  9 +++------
 3 files changed, 33 insertions(+), 11 deletions(-)

diff --git a/kernel/evl/net/dev.c b/kernel/evl/net/dev.c
index 9361b7fed9d3..4a895dc4bfbf 100644
--- a/kernel/evl/net/dev.c
+++ b/kernel/evl/net/dev.c
@@ -306,15 +306,25 @@ static void disable_oob_port(struct net_device *dev) /* inband, rtnl_lock held *
 	evl_flush_wait(&est->tx_wait, EVL_T_RMID);
 	evl_schedule();
 
-	netif_disable_oob_diversion(real_dev);
-
-	evl_stop_kthread(est->rx_handler);
-	evl_destroy_flag(&est->rx_flag);
+	/*
+	 * The order is important: first we stop the TX thread so that
+	 * the NIC driver may assume that we won't do any hard xmit
+	 * from the oob stage anymore. Then we disable diversion in
+	 * the NIC driver before stopping the RX thread, so that we
+	 * may assume that no RX won't happen from the oob stage
+	 * anymore. In addition, proper synchronization is performed
+	 * by both threads to flush the pending traffic.
+	 */
 	if (est->tx_handler) {
 		evl_stop_kthread(est->tx_handler);
 		evl_destroy_flag(&est->tx_flag);
 	}
 
+	netif_disable_oob_diversion(real_dev);
+
+	evl_stop_kthread(est->rx_handler); /* Stop before destroy the flag. */
+	evl_destroy_flag(&est->rx_flag);
+
 	__set_rx_filter(est, NULL);
 	evl_net_dev_purge_pool(real_dev);
 	/* Once the TX kthread is stopped, qdisc is idle. */
diff --git a/kernel/evl/net/input.c b/kernel/evl/net/input.c
index 0171ec8b85c0..3dd0988ca3f2 100644
--- a/kernel/evl/net/input.c
+++ b/kernel/evl/net/input.c
@@ -89,7 +89,7 @@ void evl_net_do_rx(void *arg)
 		while (!test_bit(EVL_NETDEV_RX_SCHED_BIT, &est->flags)) {
 			ret = evl_wait_flag(&est->rx_flag);
 			if (ret)
-				break;
+				goto flush_out;
 		}
 
 		/* Poll oob-capable drivers for feeding rx_packets. */
@@ -118,6 +118,21 @@ void evl_net_do_rx(void *arg)
 
 		evl_net_ipv4_gc(dev_net(dev));
 	}
+
+flush_out:
+	/*
+	 * Drop unconsumed buffers since the interface goes down,
+	 * don't even try to inject them into the stack so that we
+	 * won't allow trackers holding references on that device to
+	 * be created. Since we've passed the crossing in
+	 * disable_oob_port(), this is the only safe thing to do.
+	 */
+	if (evl_net_move_skb_queue(&est->rx_packets, &list)) {
+		list_for_each_entry_safe(skb, next, &list, list) {
+			list_del(&skb->list);
+			evl_net_free_skb(skb);
+		}
+	}
 }
 
 static inline void wake_rx(struct evl_netdev_state *est)
diff --git a/kernel/evl/net/output.c b/kernel/evl/net/output.c
index 9ca8cb90eebc..712da8144239 100644
--- a/kernel/evl/net/output.c
+++ b/kernel/evl/net/output.c
@@ -118,18 +118,13 @@ void evl_net_do_tx(void *arg)
 	int ret;
 
 	est = dev->oob_state.estate;
+	qdisc = est->qdisc;
 
 	while (!evl_kthread_should_stop()) {
 		ret = evl_wait_flag(&est->tx_flag);
 		if (ret)
 			break;
 
-		/*
-		 * FIXME: stax-protect this against swap while pulling
-		 * packets.
-		 */
-		qdisc = est->qdisc;
-
 		/*
 		 * Transmit the traffic according to the
 		 * prioritization implemented by the queueing
@@ -150,6 +145,8 @@ void evl_net_do_tx(void *arg)
 			do_tx(qdisc, dev, skb, more);
 		}
 	}
+
+	qdisc->oob_ops->flush(qdisc);
 }
 
 static void skb_xmit_inband(struct sk_buff *skb)
-- 
2.30.2

